src/psyclone/tests/psyir/frontend/fparser2_program_handler_test.py

```diff
@@ -77,3 +79,27 @@ def test_program_handler(parser):
     writer = FortranWriter()
     result = writer(psyir)
     assert result == expected
+
+
+def test_program_handler_no_module(parser):
+    ''' Test that the FileContainer contains the relevant RoutineSymbol
+    nodes when we have no containing module.'''
+    code = '''
+    subroutine a()
+    end subroutine
+    real function b(val)
+       real :: val
+       b = val
+    end function
+    '''
```

Comment:
Can you add test for the following cases (no worries if some done work - they can be xfail, I am just interested to know where we are with this complex cases) :
- Combine different type of program units:
```
program test
end program test
module A
end module A
subroutine mysub
end subroutine mysub
```
- Have non-named programs:
```
subroutine mysub(a)
   integer, intent(inout) :: a
   a = a + 1
end subroutine
end
```
This is valid fortran, note the seconds end, here mysub is not directly in the FileContainer, but in a program with no name.
Similar to the previous case but with variable declaration in the program, this is also valid Fortran:
```
integer :: b, c
contains
subroutine mysub(a)
   integer, intent(inout) :: a
   a = a + 1
end subroutine
end
```

Reply:
All of these are added as failing tests. The latter 2 produce Code blocks and the first produces the correct objects but no symbols for anything other than the subroutine

Reply:
Xfailing

Reply:
Added these now.